chemprop/train/loss_functions.py

```diff
@@ -171,16 +223,217 @@ def wasserstein_loss(model_spectra: torch.tensor, target_spectra: torch.tensor,
+        model_spectra = torch.where(
+            model_spectra < threshold, threshold_sub, model_spectra
+        )
     model_spectra = torch.where(mask, model_spectra, zero_sub)
     sum_model_spectra = torch.sum(model_spectra, axis=1, keepdim=True)
     model_spectra = torch.div(model_spectra, sum_model_spectra)
 
     # Calculate loss value
-    target_cum = torch.cumsum(target_spectra,axis=1).to(torch_device)
-    model_cum = torch.cumsum(model_spectra,axis=1).to(torch_device)
+    target_cum = torch.cumsum(target_spectra, axis=1).to(torch_device)
+    model_cum = torch.cumsum(model_spectra, axis=1).to(torch_device)
     loss = torch.abs(target_cum - model_cum)
     loss = loss.to(torch_device)
 
     return loss
+
+
+def normal_mve(pred_values, targets):
+    """
+    Use the negative log likelihood function of a normal distribution as a loss function used for making
+    simultaneous predictions of the mean and error distribution variance simultaneously.
+
+    :param pred_values: Combined predictions of means and variances of shape(data, tasks*2).
+                        Means are first in dimension 1, followed by variances.
+    :return: A tensor loss value.
+    """
+    # Unpack combined prediction values
+    pred_means, pred_var = torch.split(pred_values, pred_values.shape[1] // 2, dim=1)
+    pred_means = pred_means.to(pred_values.device)
+    pred_var = pred_var.to(pred_values.device)
+
+    return torch.log(2 * np.pi * pred_var) / 2 + (pred_means - targets) ** 2 / (
+        2 * pred_var
+    )
+
+
+# evidential classification
+def dirichlet_class_loss(alphas, target_labels, lam=0):
+    """
+    Use Evidential Learning Dirichlet loss from Sensoy et al in classification datasets.
+    :param alphas: Predicted parameters for Dirichlet in shape(datapoints, tasks*2).
+                   Negative class first then positive class in dimension 1.
+    :param target_labels: Digital labels to predict in shape(datapoints, tasks).
+    :lambda: coefficient to weight KL term
+
+    :return: Loss
+    """
+
+    def KL(alpha):
+        """
+        Compute KL for Dirichlet defined by alpha to uniform dirichlet
+        :alpha: parameters for Dirichlet
+
+        :return: KL
+        """
+        beta = torch.ones_like(alpha)
+        S_alpha = torch.sum(alpha, dim=-1, keepdim=True)
+        S_beta = torch.sum(beta, dim=-1, keepdim=True)
+
+        ln_alpha = torch.lgamma(S_alpha) - torch.sum(
+            torch.lgamma(alpha), dim=-1, keepdim=True
+        )
+        ln_beta = torch.sum(torch.lgamma(beta), dim=-1, keepdim=True) - torch.lgamma(
+            S_beta
+        )
+
+        # digamma terms
+        dg_alpha = torch.digamma(alpha)
+        dg_S_alpha = torch.digamma(S_alpha)
+
+        # KL
+        kl = (
+            ln_alpha
+            + ln_beta
+            + torch.sum((alpha - beta) * (dg_alpha - dg_S_alpha), dim=-1, keepdim=True)
+        )
+        return kl
```

Comment:
why define a temporary function that you only use once?

Reply:
This function is a copy/paste from the repo associated with the evidential uncertainty paper. I kept their form for this one, but it does need to be cleaned up.